### PR TITLE
add command to generate Rust-part CLI completions

### PR DESCRIPTION
**Pull request by bczhc**
An easy way to generate shell completions is to use Rust [`clap_complete`](https://docs.rs/clap_complete/4.4.4/clap_complete) crate. Another method is to manually write the shell completion scripts, like [`cargo`](https://github.com/rust-lang/cargo/tree/master/src/etc) does :).

This PR added
an subcommand to output these completion scripts for shells. I don't know if this approach is idiomatic, however `deno` leaves this command and I just imitate its way:
```console
~ ❯ deno --help                                                                                                             20:21:26
A modern JavaScript and TypeScript runtime
...
Usage: deno [OPTIONS] [COMMAND]

Commands:
...
  completions  Generate shell completions
```

Usage example:
```console
~/bcachefs-tools master ⇡1 ❯ ./bcachefs --help
...
Miscellaneous:
  completions              Generate shell completions
  version                  Display the version of the invoked bcachefs tool
~/bcachefs-tools master ⇡1 ❯ ./bcachefs completions zsh > ~/.zfunc/_bcachefs                                    20:26:44
~/bcachefs-tools master ⇡1 ❯ exec zsh                                                                           20:26:51
~/bcachefs-tools master ⇡1 ❯ ./bcachefs <TAB>                                                             20:26:52
completions  -- Generate shell completions
help         -- Print this message or the help of the given subcommand(s)
list         --
mount        -- Mount a bcachefs filesystem by its UUID
```

### CODE REVIEW

**Comment by bczhc** [comment]
For the code style, I can't find a proper rustfmt configuration to make struct fields/function parameters aligned. If this is a deal, help me and re-format them.

For shell completions about other `bcachefs` subcommands, there are several ways:
- Manually write shell completion rules
- Use some standalone programs/tools to translate getopt-style help messages to shell completions
- Write dummy `clap` commands and still use `clap_complete` to generate them :).

File: rust-src/src/lib.rs [hunk 1]
```diff
@@ -34,12 +34,25 @@ macro_rules! c_str {
...
-        .map(|i| ::std::ffi::OsStr::from_bytes(i.to_bytes()))
-        .collect();
+            .map(|i| unsafe { ::std::ffi::CStr::from_ptr(*$argv.add(i as usize)) })
+            .map(|x| {
+                ::cfg_if::cfg_if! {
+                    if #[cfg(unix)] {
+                        use ::std::os::unix::ffi::OsStrExt;
+                        ::std::ffi::OsStr::from_bytes(x.to_bytes())
+                    } else {
+                        // If arbitrary-bytes `OsStr`s are not supported (e.g. MS Windows),
```
Comment (koverstreet) [inline rust-src/src/lib.rs:45]: No, this is crap. We're not supporting non-unix targets any time soon, and if we do we definitely won't be dumping #ifdefs in the code like this - I'll expect this to be written with a proper helper.

**Review by koverstreet — APPROVED** [review]
This looks fine

File: rust-src/src/lib.rs (hunk 1 above)
Comment (bczhc) [reply rust-src/src/lib.rs]: Alright and yes, helper "Ext" traits can be used here. And due to non-unix targets are not considered in a short time, for now I'm just going to revert this.

**Comment by koverstreet** [comment]
This has been merged.

**Pull request closed without merging**